Projects: remake issues tab as project detail

chadoh opened the pull request:
fixes #1865

Rather than having an Issues tab at the top that, when visited, shows all issues for all projects, you can now only see issues for one project at a time.

This has a few benefits:

* A simpler filtering experience: three filters instead of four.
* Less wasted vertical space: the filter bar now contains a back button, and there's no need to show tabs. Your list of issues displays closer to the top of the page.
* Code maintainability: expect faster changes to projects, now that we were able to simplify the complex logic resulting from multi-project filtering.

I recommend viewing this diff [with indentation changes hidden](https://github.com/AutarkLabs/open-enterprise/pull/1902/files?w=1)

## Screenshot

Three quirks to note:

1. When clicking through from a project to an issue, the heading of the page disappears momentarily – this would be awkward to code around, and I don't think it's worth it. The reason this happens is because it is coded in a way that keeps the URL minimal (project detail at `/projects/:projectId`, issue detail at `/issues/:issueId`), and ensures that refreshing the page will still load data correctly and show the correct heading

2. When you apply filters, then select an issue, then click "Back", your filters are lost – this has two causes: 1. We are not currently implementing filtering via the URL, 2. Even if we were, aragonAPI does not support true "go back" behavior; we could hack something together ourselves, but it would work imperfectly

3. The "Show More" behavior has not changed – you'll see how clumsy it is for large projects in the screen recording below

![ProjectDetail 2020-01-21 at 17 19 18](https://user-images.githubusercontent.com/221614/72848253-3452c780-3c72-11ea-8220-3b612fffd08b.gif)

Commit:
feat(projects): remake issues tab as project detail

Rather than having an Issues tab at the top that, when visited, shows
all issues for all projects, you can now only see issues for one project
at a time.

This has a few benefits:

* A simpler filtering experience: three filters instead of four.
* Less wasted vertical space: the filter bar now contains a back button,
  and there's no need to show tabs. Your list of issues displays closer
  to the top of the page.
* Code maintainability: expect faster changes to projects, now that we
  were able to simplify the complex logic resulting from multi-project
  filtering

javieralaves reviewed (dismissed):
Looks safe enough to me 👍

ottodevs commented on apps/projects/app/components/Content/IssueDetail/index.js:
```diff
@@ -18,21 +18,24 @@ import DetailsCard from './DetailsCard'
       <Bar>
-        <BackButton onClick={() => requestPath('/issues')} />
+        <BackButton onClick={() => {
+          if (repo) requestPath('/projects/' + repo.data._repo)
```
`_repo` terminology here seems a bit confusing

chadoh replied:
Yes, I agree, I'd love to rethink this data structure and get rid of `_repo`! However, this is the data structure used all over our app – I think its roots might be in our script.js logic. Changing it seems far outside the scope of this change.

ottodevs commented on apps/projects/app/utils/gql-queries.js:
```diff
@@ -33,16 +33,15 @@ const issueAttributes = `
         issues(
           states:OPEN,
-          first: ${repos[repoId].fetch},
-          ${repos[repoId].showMore ? `after: "${repos[repoId].endCursor}",` : ''}
```
should we want to keep tracking cursor for multi-page issues results?

In case we don't need that anymore, it does not make sense to keep using the "show more results" button as we don't have any mechanism to fetch more items

chadoh replied:
We do!

See the behavior of the "Show More" button:

```js
onClick={() => {
  this.setState({ cachedIssues: allIssues })
  this.props.setQuery({
    after: data.repository.issues.pageInfo.endCursor,
  })
}}
```

This `setQuery` changes the query params in `ProjectDetailWrap`, setting an `after` attribute to the `endCursor` from the current query. This causes `ProjectDetailQuery` to fire a new query, which gets passed down to `ProjectDetail` as `graphqlQuery`. Then `ProjectDetail` has access to this new `data` object, which in turn has a new `data.repository.issues.pageInfo.endCursor`

ottodevs commented on apps/projects/app/components/Shared/FilterBar/FilterBar.js:
```diff
@@ -614,30 +591,14 @@ const FilterMenuItem = styled(ContextMenuItem)`
   display: flex;
+  margin-left: ${GU}px;
   > * {
     margin-right: 8px;
```
Shouldn't be too much to update this to use `GU` in the same run

ottodevs reviewed (dismissed):
Thanks @chadoh great work going on here.

I like how this PR cleans out a lot of code and complexity, I like the projects codebase a lot more in this new state, I am still a bit concerned that removing multi-projects filtering feature puts us a bit behind projects like ZenHub where that works seamlessly, but I guess is matter of trade-offs and definitely rather this approach for the time being.

The limitations you commented on definitely seem to have proper solutions, but I agree for now it is reasonable to separate into future tasks.

The left-margin between back button and first filter button seems a bit tight for me.

I added some minor comment that not necessarily need further changes

The filterbar seems completely unusable at this point, the worst iteration I experienced on it... up to 10 seconds just to open a filter, the counts are not accurate and the whole app completely freezes on the Project details view... I don't think it should be a requirement to use a NASA computer to filter some issues so please, let's take a look to that.

Also there are lot of warnings and errors in the browser console, maybe those are out of scope, but we should be fixing them progressively why we implement new features.

I approve the PR because from the code perspective and acceptance criteria it looks good for me, but I am deeply concerned about these performance issues I faced with this new view, that supposedly should work just fine.

Commit:
refactor: remove unused 'query' logic from usePathHelpers

Commit:
refactor(projects): move hook out of Issue card

Since we currently render many (100+) Issue cards in this list, we want
to make sure each one renders as quickly as possible. One way we can do
this is to reduce the number of hooks this component calls.

This makes it so the `usePathHelpers` hook is processed once, rather
than 100+ times.

chadoh commented:
@ottodevs good points on the performance issues. I've added two `refactor` commits here that address it somewhat, but it seems that the biggest reason the performance has regressed is the inclusion of `react-axe`. If you comment out the block that adds it in `index.js`, the performance now seems to match the performance on Rinkeby & mainnet.
